### PR TITLE
Fix bigint PHP_INT_MIN/PHP_INT_MAX string to int convert

### PR DESCRIPTION
**Pull request by mvorisek**
|      Q       |   A
|------------- | -----------
| Type         | bug
| Fixed issues | n/a

#### Summary

Resolve https://github.com/doctrine/dbal/pull/6177#discussion_r1613456751 discussion and related original #6177.

Whole native php `int` range is guaranteed to be supported per https://www.doctrine-project.org/projects/doctrine-dbal/en/4.0/reference/types.html#bigint docs.

### CODE REVIEW

**Comment by mvorisek** [comment]
@derrabus can you please review?

**Comment by derrabus** [comment]
Please don't ping me on PRs please.

**Comment by mvorisek** [comment]
I pinged you because of you authored the original PR, sorry.

File: src/Types/BigIntType.php [hunk 1]
```diff
@@ -47,15 +48,21 @@ public function convertToPHPValue(mixed $value, AbstractPlatform $platform): int
...
-        }
-
         assert(
             is_string($value),
             'DBAL assumes values outside of the integer range to be returned as string by the database driver.',
         );
 
+        // workaround https://github.com/php/php-src/issues/14345
+        $dotPos = strpos($value, '.');
+        if ($dotPos !== false && rtrim(substr($value, $dotPos + 1), '0') === '') {
```
Comment (mvorisek) [inline src/Types/BigIntType.php:58]: Psalm issue - https://github.com/vimeo/psalm/issues/10995

**Comment by derrabus** [comment]
Please keep in mind that this piece of code will be executed on a hot path. If I hydrate thousands of entities with bigint fields, I don't want to execute `preg_match()` for each of them. Your implementation is too expensive.

**Comment by mvorisek** [comment]
> Please keep in mind that this piece of code will be executed on a hot path. If I hydrate thousands of entities with bigint fields, I don't want to execute `preg_match()` for each of them. Your implementation is too expensive.

PCRE JIT is very fast, but yes, the regex replace is possible to be coded /wo regex and I alredy considered that option because only limited number (4096) are cached. I will rework the code.

File: src/Types/BigIntType.php [hunk 2]
```diff
@@ -47,15 +49,35 @@ public function convertToPHPValue(mixed $value, AbstractPlatform $platform): int
...
-        if ($value > PHP_INT_MIN && $value < PHP_INT_MAX) {
-            return (int) $value;
-        }
-
         assert(
             is_string($value),
             'DBAL assumes values outside of the integer range to be returned as string by the database driver.',
         );
 
+        if (str_starts_with($value, '-') || str_starts_with($value, '+')) {
```
Comment (mvorisek) [inline src/Types/BigIntType.php:57]: Psalm is broken - https://github.com/vimeo/psalm/issues/10995 - how to fix the CI?

**Comment by mvorisek** [comment]
> Please keep in mind that this piece of code will be executed on a hot path. If I hydrate thousands of entities with bigint fields, I don't want to execute `preg_match()` for each of them. Your implementation is too expensive.

done

**Comment by derrabus** [comment]
Sorry, but that implementation is way too complicated. I don't want to maintain this.

**Comment by mvorisek** [comment]
The implementation is minimal, we strip leading plus sign, zeros and trailing zeros after decimal point. If the number is then castable into int without precision loss, we cast it.

The leading/trailing zeros should be stripped because the input number can come from sources with explicit digits/decimal configured. This behaviour is tested and I do not think it can be implemented simpler. If you have an idea how to implemenet this simpler/better, I am of course ready for your ideas.

**Comment by mvorisek** [comment]
If leading/trailing zeros should be supported, the impl. is minimal IMO. I would be happy if this can either be merged as is or please let me know how to fix the min/max issue differently.

**Comment by derrabus** [comment]
Which DBMS formats 2^31-1 with leading zeros?

**Comment by mvorisek** [comment]
I tested all DBs using https://dbfiddle.uk/6OSky-ka and none DB vendor prepend leading zeros even for DECIMAL type by default.

So you want me to remove the "leading zeros accepting" code in order to save a few lines of code?

**Comment by derrabus** [comment]
🤷‍♂️

**Comment by mvorisek** [comment]
I am asking as the integer DBAL Type class can be used to load any value, the value might some from user data, formatted string column, ...

**Comment by derrabus** [comment]
> I am asking as the integer DBAL Type class can be used to load any value

Sure, but it's not meant to be used on any value. And I refuse to build workarounds to accommodate misuse.

File: src/Types/BigIntType.php [hunk 3]
```diff
@@ -47,15 +48,24 @@ public function convertToPHPValue(mixed $value, AbstractPlatform $platform): int
...
-            return (int) $value;
-        }
-
         assert(
             is_string($value),
             'DBAL assumes values outside of the integer range to be returned as string by the database driver.',
         );
 
+        $dotPos = strpos($value, '.');
+        if ($dotPos !== false && rtrim(substr($value, $dotPos + 1), '0') === '') {
```
Comment (mvorisek) [inline src/Types/BigIntType.php:57]: Psalm is broken - https://github.com/vimeo/psalm/issues/10995 - how to fix the CI?

**Comment by mvorisek** [comment]
> > I am asking as the integer DBAL Type class can be used to load any value
> 
> Sure, but it's not meant to be used on any value. And I refuse to build workarounds to accommodate misuse.

simplified in https://github.com/doctrine/dbal/pull/6410/commits/2cc0fae35310d65b013b65f68e6aa05e40ecb048

File: src/Types/BigIntType.php [hunk 4]
```diff
@@ -47,15 +48,24 @@ public function convertToPHPValue(mixed $value, AbstractPlatform $platform): int
...
-        if ($value > PHP_INT_MIN && $value < PHP_INT_MAX) {
-            return (int) $value;
-        }
-
         assert(
             is_string($value),
             'DBAL assumes values outside of the integer range to be returned as string by the database driver.',
         );
 
+        $dotPos = strpos($value, '.');
```
Comment (derrabus) [inline src/Types/BigIntType.php]: Why do we expect a dot in an integer value coming from the database?

Comment (mvorisek) [reply src/Types/BigIntType.php]: see https://dbfiddle.uk/6OSky-ka - if database column type (ex. `decimal`) does not match the used type (ex. `bigint`) the value must be parsed correctly

Comment (derrabus) [reply src/Types/BigIntType.php]: No, it must not. See https://github.com/doctrine/dbal/pull/6410#issuecomment-2171891186

File: src/Types/BigIntType.php [hunk 5]
```diff
@@ -47,15 +46,19 @@ public function convertToPHPValue(mixed $value, AbstractPlatform $platform): int
...
-        if ($value > PHP_INT_MIN && $value < PHP_INT_MAX) {
-            return (int) $value;
-        }
-
         assert(
             is_string($value),
             'DBAL assumes values outside of the integer range to be returned as string by the database driver.',
         );
 
+        if (str_starts_with($value, '+') || $value === '-0') {
```
Comment (derrabus) [inline src/Types/BigIntType.php]: You do realize what I'm about to ask, don't you. 🙃

Comment (mvorisek) [reply src/Types/BigIntType.php]: I have removed even this, although I have done it only for you, my proposal is this PR without the last 3 commits as IMO the type should accept well formed integer string. All robust libs do that and in my company we use DBAL Types to cast data even from APIs for example. If you however persist on minimalistic impl., I am ok to give my code away even if I am not happy with it.

File: src/Types/BigIntType.php [hunk 6]
```diff
@@ -47,15 +44,15 @@ public function convertToPHPValue(mixed $value, AbstractPlatform $platform): int
...
-            return (int) $value;
-        }
-
         assert(
             is_string($value),
             'DBAL assumes values outside of the integer range to be returned as string by the database driver.',
         );
 
+        if ($value === (string) (int) $value) {
+            return (int) $value;
```
Comment (mvorisek) [inline src/Types/BigIntType.php:59]: psalm is wrong here - https://github.com/vimeo/psalm/issues/10995

tell me how to fix the CI

File: src/Types/BigIntType.php (hunk 5 above)
Comment (derrabus) [reply src/Types/BigIntType.php]: I'm going to answer this in the main thread.

**Comment by derrabus** [comment]
> in my company we use DBAL Types to cast data even from APIs for example

But that's an off-label use and I hope you're fully aware of that. If your goal is to normalize data coming from arbitrary external sources, there are libraries tailored to that task, like Symfony Serializer for example. DBAL does neither want nor need to compete with such libraries. If our type conversion system does not work for your clearly undocumented use-case, please use something else. Honestly.

> If you however persist on minimalistic impl., I am ok to give my code away even if I am not happy with it.

Okay, I'd like to settle this once and for all. I really appreciate the dedication with which you contribute to our libraries. I really do. However, we want to keep our libraries focused and maintainable. And your initial proposal was anything but focused and maintainable. You've proposed to merge a piece of code that nobody but you will ever need and which at the same time would significantly slow down our type conversion for everybody else.

The time that we maintainers can spend on open source work is finite and mostly unpaid. The time that I waste on unnecessary discussions – like this one – is everybody's loss because I cannot spend it on reviewing other contributions, yours included.

File: src/Types/BigIntType.php [hunk 7]
```diff
@@ -47,15 +44,15 @@ public function convertToPHPValue(mixed $value, AbstractPlatform $platform): int
             return $value;
         }
 
-        if ($value > PHP_INT_MIN && $value < PHP_INT_MAX) {
-            return (int) $value;
-        }
```
Comment (derrabus) [inline src/Types/BigIntType.php:52]: Please keep this check. It will find the most common cases and should be faster than casting back and forth.

Comment (mvorisek) [reply src/Types/BigIntType.php:52]: done

**Comment by greg0ire** [comment]
I don't think this is going anywhere, sorry.

**Comment by mvorisek** [comment]
@greg0ire what you you mean, there is a bug and I coded the tests and the fix, please reopen this PR and let me know what changes, if any, you want.

**Comment by mvorisek** [comment]
@greg0ire with Alexander I got to a point when this PR is lighweight and passing the tests. Can you please clarify "I don't think this is going anywhere" into a feedback I can act on? Can you please reopen this PR?

**Comment by greg0ire** [comment]
Clarification: I don't think there is anything you can do.

**Comment by mvorisek** [comment]
Sorry, do you understand the problem, are you aware that valid max. int `2^64-1` (and min.) value is currently not properly casted to `int` type?

**Comment by greg0ire** [comment]
I don't understand the problem, and I do not see an explanation about it anywhere.

**Comment by derrabus** [comment]
> I don't understand the problem

The problem ist that if the DB would return PHP_INT_MAX as string, we would not convert it to int although one might expect that.

**Comment by greg0ire** [comment]
I won't block this anymore but I don't get the fix, feel free to deal with this if you understand it @derrabus

File: src/Types/BigIntType.php (hunk 6 above)
Comment (mvorisek) [reply src/Types/BigIntType.php:59]: @greg0ire any advice how to fix the psalm CI?

Comment (mvorisek) [reply src/Types/BigIntType.php:59]: @greg0ire how should I make psalm happy when there is known bug in psalm?

File: tests/Functional/Types/BigIntTypeTest.php [hunk 8]
```diff
@@ -30,13 +28,13 @@ public function testSelectBigInt(string $sqlLiteral, int|string|null $expectedVa
 
         $this->connection->executeStatement(<<<SQL
             INSERT INTO bigint_type_test (id, my_integer)
-            VALUES (42, $sqlLiteral)
+            VALUES (1, $sqlLiteral)
```
Comment (derrabus) [inline tests/Functional/Types/BigIntTypeTest.php]: What is the reason for this change?

File: src/Types/BigIntType.php [hunk 9]
```diff
@@ -47,15 +47,18 @@ public function convertToPHPValue(mixed $value, AbstractPlatform $platform): int
...
-
         assert(
             is_string($value),
             'DBAL assumes values outside of the integer range to be returned as string by the database driver.',
         );
 
+        if (
+            ($value > PHP_INT_MIN && $value < PHP_INT_MAX)
+            || $value === (string) (int) $value
```
Comment (derrabus) [inline src/Types/BigIntType.php:57]: I think, we can still get rid of these casts. There are precisely four cases in which our check `$value > PHP_INT_MIN && $value < PHP_INT_MAX` prevents an integer cast. We might as well just enumerate them.

File: src/Types/BigIntType.php (hunk 6 above)
Comment (derrabus) [reply src/Types/BigIntType.php:59]: Ignoring the error is fine.

File: src/Types/BigIntType.php (hunk 9 above)
Comment (mvorisek) [reply src/Types/BigIntType.php:57]: Do you mean x86 & x64 min/max?

I wonder why would someone prefer hardcoding values?

TBH, I actually think `$value > PHP_INT_MIN && $value < PHP_INT_MAX` is not ideal as it rounds currently.

File: tests/Functional/Types/BigIntTypeTest.php (hunk 8 above)
Comment (mvorisek) [reply tests/Functional/Types/BigIntTypeTest.php]: Do you want to keep `42` over `1`?

See the whole tests changes, I simplified it, no need to use `where` - thus any number/ID can be here.

Comment (derrabus) [reply tests/Functional/Types/BigIntTypeTest.php]: So, the change was unrelated and unnecessary? Then, let's revert.

**Comment by derrabus** [comment]
As soon as the unnecessary changes to the test have been reverted, I'm fine with the PR.

File: src/Types/BigIntType.php (hunk 9 above)
Comment (derrabus) [reply src/Types/BigIntType.php:57]: Very well, lets not over-optimize this.

File: tests/Functional/Types/BigIntTypeTest.php (hunk 8 above)
Comment (mvorisek) [reply tests/Functional/Types/BigIntTypeTest.php]: done in https://github.com/doctrine/dbal/pull/6410/commits/8425fac079f01dce56f6885c0307ab6591df8efe - is that really helpful?

The original `testSelectBigIntEdge` is useless as the value is inserted as string and thus there is no conversion thru the DBAL and thus it does not cover anything over `testSelectBigInt`.

**Review by derrabus — APPROVED** [review]
